### PR TITLE
Man pages

### LINKED ISSUE #208: make a better / real man page
README.Debian from the debian package tells about quite some difficulties about creating a man page.

I think this should be solved upstream (== by us), also to make it easier for other linux distributions.

Debian packaging is discussed here:

https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1355849.html

The solution for creating the manpage should be automated.

---

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/26801934-make-a-real-man-page)

### PR DESCRIPTION
**Pull request by enkore**
Fixes #208 

This isn't the pinnacle of perfection yet (and I don't believe in first-time perfection anyway ;), some things to improve:

- There are no overview pages yet, like `borg(1)`.
- ~~No examples in the pages (but they aren't in --help either, maybe kill two birds with one stone here)~~
- ~~SEE ALSO isn't fleshed out yet, only a stub (borg-common(1)) + the infrastructure to do them. borg-create(1) has some.~~

However, in general they look quite good and read well.

Code wise there is some duplication with the build_usage code, but it's more or less different all over the place so deduping that would make it harder to understand or inflate the code more.

### CODE REVIEW

**Comment by enkore** [comment]
Since I don't know a way to display fancy HTML from man pages on or offline, here are two Konsole-grabs in lieu of that: https://pstore.enkore.de/borgman.html

**Comment by enkore** [comment]
A page with more complex layout http://pstore.enkore.de/borgman2.html

File: setup.py [hunk 1]
```diff
@@ -326,6 +329,222 @@ def shipout(text):
...
+        pass
+
+    def finalize_options(self):
+        pass
+
+    def run(self):
+        try:
+            import docutils
+        except ImportError:
+            print('Cannot generate man pages -- docutils is not installed.', file=sys.stderr)
```
Comment (ThomasWaldmann) [inline setup.py]: this is dangerous, it hides information and might tell wrong conclusions at some times, e.g. if there is an ImportError inside docutils.

i'ld prefer not catching this.

File: setup.py [hunk 2]
```diff
@@ -326,6 +329,222 @@ def shipout(text):
...
+            return
+
+        for command, parser in sorted(choices.items()):
+            if command.startswith('debug') or command == 'help':
+                continue
+
+            man_title = 'borg-' + command.replace(' ', '-')
+            print('building man page', man_title + '(1)', file=sys.stderr)
+
+            if self.generate_level(command + " ", parser, Archiver):
```
Comment (ThomasWaldmann) [inline setup.py]: always use single quotes for (short) string literals?

File: setup.py [hunk 3]
```diff
@@ -326,6 +329,222 @@ def shipout(text):
...
+    def write_examples(self, write, command):
+        with open('docs/usage.rst') as fd:
+            usage = fd.read()
+            usage_include = '.. include:: usage/%s.rst.inc' % command
+            begin = usage.find(usage_include)
+            end = usage.find('.. include', begin + 1)
+            examples = usage[begin:end]
+            examples = examples.replace(usage_include, '')
+            examples = examples.replace('Examples\n~~~~~~~~', '')
+            examples = examples.replace('Miscellaneous Help\n------------------', '')
```
Comment (ThomasWaldmann) [inline setup.py]: there could be a underlined(text, c) function that returns 'text\ncccc\n'.

File: setup.py [hunk 4]
```diff
@@ -326,6 +329,222 @@ def shipout(text):
...
+        write(', '.join(see_also))
+
+    def gen_man_page(self, name, rst):
+        from docutils.writers import manpage
+        from docutils.core import publish_string
+        man_page = publish_string(source=rst, writer=manpage.Writer())
+        with open('docs/man/%s.1' % name, 'wb') as fd:
+            fd.write(man_page)
+
+    def write_usage(self, parser, fp):
```
Comment (ThomasWaldmann) [inline setup.py]: use fd, as elsewhere?

File: src/borg/archiver.py [hunk 5]
```diff
@@ -2024,16 +2023,17 @@ def build_parser(self, prog=None):
...
         archive_group.add_argument('-c', '--checkpoint-interval', dest='checkpoint_interval',
                                    type=int, default=1800, metavar='SECONDS',
                                    help='write checkpoint every SECONDS seconds (Default: 1800)')
         archive_group.add_argument('--chunker-params', dest='chunker_params',
                                    type=ChunkerParams, default=CHUNKER_PARAMS,
-                                   metavar='CHUNK_MIN_EXP,CHUNK_MAX_EXP,HASH_MASK_BITS,HASH_WINDOW_SIZE',
-                                   help='specify the chunker parameters. default: %d,%d,%d,%d' % CHUNKER_PARAMS)
+                                   metavar='PARAMS',
+                                   help='specify the chunker parameters (CHUNK_MIN_EXP, CHUNK_MAX_EXP, '
+                                        'HASH_MASK_BITS, HASH_WINDOW_SIZE). default: %d,%d,%d,%d' % CHUNKER_PARAMS)
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: add a \n in here, this is longer than 80 chars.

Comment (ThomasWaldmann) [inline src/borg/archiver.py]: also check other places for len > 80.

Comment (enkore) [inline src/borg/archiver.py]: we use 120 chars line length

File: setup.py (hunk 3 above)
Comment (enkore) [inline setup.py]: Since this removes verbatim text from the usage.rst I'd rather keep it obvious/verbatim.

File: src/borg/archiver.py (hunk 5 above)
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: i didn't mean the src formatting here, i meant the console output.

Comment (enkore) [inline src/borg/archiver.py]: I believe parameter help is wrapped by argparse

**Pull request merged**